### DISCUSSION TITLE
Strange behavior of ESP_IR_TR with SerBr Tx or IRsend on GPIO4/GPIO14

### DISCUSSION THREAD

**Discussion by TheChatty** [arendst/Tasmota#16447] [Ideas]
### PROBLEM DESCRIPTION
I own an [ESP_IR_TR](https://ae01.alicdn.com/kf/S7c37a1d55ab545909dca7446ff60997cZ/ESP8285-ESP-01M-IR-Transceiver-Wifi-Wireless-Module-Remote-Control-Switch-Development-Learning-Board-esp-8285.jpg_Q90.jpg_.webp) which connects IR emitter to GPIO4 and IR receiver to GPIO14.

When configured to **IRsend/IRrecv** it **works** like it should: emitter usually dark, receives commands from real TV remote, sends those commands to real TV.

When configured to **SerBr Tx/Rx** the emitter is **always lit**, nothing is received when other IR head tries to send something (both configured to e.g. 9600,8n1), steady "garbage" is received by other IR head.

FYI: [other IR head](https://i.ebayimg.com/images/g/WNUAAOSwYI9h0Loa/s-l1600.png) + FTDI (working flawlessly in other environments)

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): ESP_IR_TR
- [x] Tasmota binary firmware version number used: 12.1.1.1
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: esptool.py
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
at ESP_IR_TR
```
backlog SBaudrate 9600;SSerialConfig 8N1
SSerialSend5 1F 2E 3D 4C 5B 6A 79 88 97 A6 B5 C4 D3 E2 F1
```

at other IR head:
```
SerialReceived 70 C2 85 FF C6 00 FE C2 0D FF EF D1 FE E2 00 FE F7 59 FE FB 18 00
```

### EXPECTED BEHAVIOUR
at other IR head:
```
SerialReceived 1F 2E 3D 4C 5B 6A 79 88 97 A6 B5 C4 D3 E2 F1
```

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by s-hadinger** [comment]
If I understand well you are asking for inverted IrTx which is not supported by TasmotaSerial. I'm hesitant to change TasmotaSerial driver for esp8266 since it's a delicate piece of code.

**Comment by TheChatty** [comment]
Although FTDI does not idle on HIGH [Wikipedia](https://en.wikipedia.org/wiki/Universal_asynchronous_receiver-transmitter) says: _The idle, no data state is high-voltage, or powered. This is a historic legacy from telegraphy, in which the line is held high to show that the line and transmitter are not damaged._

But main question is: **why am I not receiving what was send** with SerBr? IRsend/IRrecv was only mentioned to tell you that the hardware is working.

**Reply by s-hadinger** [reply]
If the serial signal is inverted  it is normal to receive garbage. How did you connect the IR to the FTDI?

**Reply by TheChatty** [reply]
GND/TX/RX/VCC directly to FTDI pins. I can communicate with two smart meters this way.

**Reply by s-hadinger** [reply]
Because I suppose that the IR serial transmitter inverts the signal if it's designed to be connected to a serial signal. 

Those are two different devices, right?

**Comment by s-hadinger** [comment]
Serial is always high when idle and low to mark start and end bits. This is the contrary of IR.

**Comment by barbudor** [comment]
The fact that the IR led is lit on gpio high or low depends on how it is connected to the GPIO : towards VCC or towards GND, through a transistor or not.

When using your FTDI with another IR beamer, that one may be wired differently than the ESP_IR_TR
So if you want the same behavior you need the ESP_IR_TR to match the same wiring as the other IR head

As s-hadinger said, UART standard is to have idle Hi. It makes no sense to invert that in software.

**Comment by TheChatty** [comment]
Well, my final **goal is to read two smart meters** (_Kamstrup [Multical 403](https://www.kamstrup.com/de-de/waermezaehlerloesungen/waermezaehler/meters/multical-403) (1200@8N2) and [EasyMeter Q3A](https://www.easymeter.com/products/zaehler/q3a) (9600@7E1)_) and publish their readings to MQTT.

This is **already possible with ESP-01 hardware serial and [D0-Head](https://wiki.volkszaehler.org/hardware/controllers/ir-schreib-lesekopf-ttl-ausgang)** (I can't read the schematics whether it includes an inverter but it obviously does).

**ESP-IR-TR** comes **preassembled** from China but for a different use case (IR remote commands). It includes already ESP-01M, IR emitter and IR receiver **at a friction of the cost**.

There are two things to overcome:
* [invert SerBr Tx](https://github.com/arendst/Tasmota/issues/7838#issuecomment-857095567)<br>➢ otherwise emitter needs to be unsoldered, enhanced with inverter and some reconnected to the board (probably ugly)
* [enable odd SerialConfigs](https://github.com/arendst/Tasmota/pull/12018) for software serial<br>➢ otherwise GPIO4/14 needs to shorted with GPIO1/3 to utilize hardware serial, this would also disable usage of the dupont connector for flashing etc.

**Reply by barbudor** [reply]
Switching to hardware serial won't help as it enforce the standard that UART idle is high.
If there is a way of doing it, it's only on software uart. But as s-hadinger pointed this is very timing sensitive piece of code

**Reply by TheChatty** [reply]
As I said: the goal is to **use the preassembled ESP_IR_TR without hardware modification.** The IR emitter/receiver is already connected to GPIO4/14 thus software serial.

Since we're talking about 1200 and 9600 baud I hope it won't proove as to be too timing sensitive.
